src/qibo/derivative.py

```diff
@@ -1,7 +1,180 @@
+        self.nparams = len(trainablep)
+        self.lambdaf = func
+
+    def _apply_func(self, fixed_params=None):
+        """Applies lambda function and returns final gate parameter"""
+        params = []
+        if self._featurep is not None:
+            params.append(self._featurep)
+        if fixed_params:
+            params.extend(fixed_params)
+        else:
+            params.extend(self._trainablep)
+        return self.lambdaf(*params)
+
+    def _update_params(self, trainablep=None, feature=None):
+        """Update gate trainable parameter and feature values"""
+        if trainablep:
+            self._trainablep = trainablep
+        if feature:
+            self._featurep = feature
+
+    def get_params(self, trainablep=None, feature=None):
+        """Update values with trainable parameter and calculate current gate parameter"""
+        self._update_params(trainablep=trainablep, feature=feature)
+        return self._apply_func()
+
+    def get_indices(self, start_index):
+        """Return list of respective indices of trainable parameters within
+        the optimizer's trainable parameter list"""
+        return [start_index + i for i in range(self.nparams)]
+
+    def get_fixed_part(self, trainablep_idx):
+        """Retrieve parameter constant unaffected by a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = self._trainablep[trainablep_idx]
+        return self._apply_func(fixed_params=params)
+
+    def get_scaling_factor(self, trainablep_idx):
+        """Get scaling factor multiplying a specific trainable parameter"""
+        params = [0] * self.nparams
+        params[trainablep_idx] = 1.0
+        return self._apply_func(fixed_params=params)
+
+
+def create_hamiltonian(qubit, nqubit, backend):
+    """
+    Creates appropriate Hamiltonian for a given list of qubits
+    Args:
+        qubit: qubit numbers whose states we are interested in
+        nqubit: total number of qubits, which determines size of Hamiltonian
+    Return:
+        hamiltonian: SymbolicHamiltonian
+    """
+    if not isinstance(qubit, list):
+        qubit = [qubit]
+
+    hams = []
+
+    for i in range(nqubit):
+        if i in qubit:
+            hams.append(Z(i))
+        else:
+            hams.append(I(i))
+
+    # create Hamiltonian
+    obs = np.prod([Z(i) for i in range(1)])
+    hamiltonian = SymbolicHamiltonian(obs, backend=backend)
+
+    return hamiltonian
+
+
+def error_mitigation(circuit, hamiltonian, backend, noise_model):
+    calibration = calibration_matrix(nqubits=1, backend=backend, nshots=10000)
+
+    _, _, optimal_params, _ = CDR(
+        circuit=circuit,
+        observable=hamiltonian,
+        backend=backend,
+        nshots=10000,
+        noise_model=noise_model,
+        full_output=True,
+        readout={"calibration_matrix": calibration},
+    )
+
+    return optimal_params
+
+
+def calculate_gradients(optimizer, cdr_params):
```

Comment:
The `optimizers` module needs the `derivative` one in some cases (e.g. if the optimization is a gradient descent), but I think the `derivative` should be free from this kind of dependence. 
The idea with the first `parameter_shift` rule was to pass the model (the circuit) as argument and the target parameter index. I understand here you are doing something like that, embedding into the optimizer the model's information. This is ambiguous in some way: ideally we should divide the info (1. model's info, with parameters and so on, 2. optimizer info, with the method chosen to compute gradients if needed, 3. the algorithms used to compute gradients).

I think we can implement something similar we did in the `parameter_shift` but more general, in which the `Parameter` lambda function is in some way encoded into the variational quantum circuit, which can be a `qibo.Circuit` or something more complex (a new qml_model class which can also provide the user with some pre/post-processing tools, etc).

Reply:
There exists a draft for a VQC class, however this would be implemented in a separate Pull Request